ssi-crypto/src/verification.rs

```diff
@@ -0,0 +1,340 @@
+use std::{
+    fmt,
+    ops::{BitAnd, BitAndAssign, BitOr, BitOrAssign},
+};
+
+use async_trait::async_trait;
+use iref::{Iri, IriBuf};
+use static_iref::iri;
+
+#[derive(Debug, thiserror::Error)]
+pub enum VerificationError {
+    /// Invalid proof.
+    #[error("invalid proof")]
+    InvalidProof,
+
+    /// Unsupported key identifier.
+    #[error("unsupported key id `{0}`")]
+    UnsupportedKeyId(String),
+
+    /// Invalid key identifier.
+    #[error("invalid key id `{0}`")]
+    InvalidKeyId(String),
+
+    /// Invalid key.
+    #[error("invalid key")]
+    InvalidKey,
+
+    /// Cryptographic key is not used correctly.
+    #[error("invalid use of key with <{0}>")]
+    InvalidKeyUse(ProofPurpose),
+
+    #[error("invalid signature")]
+    InvalidSignature,
+
+    /// Key controller was not found.
+    #[error("key controller `{0}` not found")]
+    KeyControllerNotFound(String),
+
+    /// Key controller is invalid.
+    #[error("invalid key controller")]
+    InvalidKeyController,
+
+    /// Unsupported controller scheme.
+    #[error("unsupported key controller scheme `{0}`")]
+    UnsupportedControllerScheme(String),
+
+    /// Verifier internal error.
+    #[error("internal error: {0}")]
+    InternalError(Box<dyn Send + std::error::Error>),
+}
+
+/// Verifier.
+#[async_trait]
+pub trait Verifier<M>: Sync {
+    /// Verify the given `signature`, signed using the given `algorithm`,
+    /// against the input `signing_bytes`.
+    async fn verify(
+        &self,
+        method: &M,
+        proof_purpose: ProofPurpose,
+        signing_bytes: &[u8],
+        signature: &[u8],
+    ) -> Result<bool, VerificationError>;
+}
+
+macro_rules! proof_purposes {
+    ($($(#[$doc:meta])* $id:ident: $variant:ident = $iri:expr),*) => {
+        /// Proof purposes.
+        #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+        pub enum ProofPurpose {
+            $(
+                $(#[$doc])*
+                $variant
+            ),*
+        }
+
+        impl ProofPurpose {
+            pub fn from_iri(iri: Iri) -> Option<Self> {
+                $(
+                    if iri == $iri {
+                        return Some(Self::$variant)
+                    }
+                )*
+
+                None
+            }
+
+            pub fn iri(&self) -> Iri<'static> {
+                match self {
+                    $(
+                        Self::$variant => $iri
+                    ),*
+                }
+            }
+        }
+
+        /// Set of proof purposes.
+        #[derive(Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+        pub struct ProofPurposes {
```

Comment:
The `ProofPurposes` type represents a set of `ProofPurpose` and expands to something like
```rust
struct ProofPurposes {
    pub assertion_method: bool,
    pub authentication: bool
    pub capability_invocation: bool,
    pub capability_delegation: bool,
    pub key_agreement: bool
}
```
Unlike a `Vec<ProofPurpose>` it does not require a memory allocation, it implements `Copy`, cannot include duplicates, provides a simple O(1) way to test included members (`p.assertion_method` instead of `p.contains(&ProofPurpose::AssertionMethod)`), and set operations (union, intersection, ...) using bit operators.

Reply:
Ah right there can't be duplicates, but then why complicate things instead of having 
```rust
struct ProofPurposes {
  assertion_method: Option<AssertionMethod>,
...
}
```

Which I realise is basically `VerificationRelationships`? I guess I should just wait to see an example/test, but I'm afraid this is over complicated for minimal benefits

Reply:
You can find a full example that shows how to build `ProofPurposes` using bit or operations here: https://github.com/spruceid/ssi/blob/c90d8c209f8938f3bb2ea34b2c199fd5b8d802d4/ssi-vc-ldp/examples/sign.rs#L73